spark/src/test/java/org/apache/iceberg/spark/source/TestParquetWrite.java

```diff
@@ -338,4 +339,65 @@ public void testPartitionedCreateWithTargetFileSizeViaOption() throws IOExceptio
     Assert.assertEquals("Should have 8 DataFiles", 8, files.size());
     Assert.assertTrue("All DataFiles contain 1000 rows", files.stream().allMatch(d -> d.recordCount() == 1000));
   }
+
+  @Test
+  public void testNestedPartitioning() throws IOException {
+    Schema nestedSchema = new Schema(
+        optional(1, "id", Types.IntegerType.get()),
+        optional(2, "data", Types.StringType.get()),
+        optional(3, "nestedData", Types.StructType.of(
+            optional(4, "id", Types.IntegerType.get()),
+            optional(5, "moreData", Types.StringType.get())))
+    );
+
+    File parent = temp.newFolder("parquet");
+    File location = new File(parent, "test");
+
+    HadoopTables tables = new HadoopTables(new Configuration());
+    PartitionSpec spec = PartitionSpec.builderFor(nestedSchema)
+        .identity("id")
+        .identity("nestedData.moreData")
```

Comment:
I'm wondering whether whether we should also fix for partition field names and not allow dots in them. Partition spec schema is used in `DataFile` schema which is used by the `ManifestReader` . 
Iceberg schemas support `indexByName` api  which maps a source column name to its field id. The column name is canonical, starting from root and containing all parent field names separated by dot.

@rdblue do you think this could somehow break the Schema `indexbyName` api?

Reply:
This dot issue in column names is fixed on the writer side by:

- Issue: https://github.com/apache/incubator-iceberg/issues/575
- PR: https://github.com/apache/incubator-iceberg/pull/220

I would say that to be consistent we should unify this on both writer and reader - that is what this PR tries to do - if we allow dots on write the reader should also work.

Reply:
Dots are allowed in all names. This is possible because we look up columns using indexByName, which has the nested representation of all field names, like "location.lat".

To build the lookup index, field names in the schema are converted once to the fully-nested name. For example, `{ "location": {"lat": ..., "long": ...}, "id" }` has 3 leaf names: location.lat, location.long, and id. Note that "location.lat" could be a top-level field with a dot in the name. That's allowed as long as there isn't a conflicting nested field. If there were a conflict, then "location.lat" would refer to two fields and building the index fails.

Using this index, we can always get the original field ID without ever needing to parse a field name using dot and traverse levels.